File: batch-failed-input-validation/python/openai_batch_validation_audit.py
"""Report OpenAI batches that failed input validation, and the lines that broke.

Read only. Two GET endpoints, /v1/batches and /v1/files, and nothing else. No
file is uploaded, no batch is created, and no failed batch is re-submitted:
re-running rows spends money on inference and only you know whether the rows
are still wanted.

On the Batch API, status "failed" is a specific claim. It means the input file
did not survive validation, which happens before any request reaches the model.
So request_counts is all zeros and nothing was billed. Individual requests that
failed inside a batch that ran are a different status and a different note.

The second half reads the file list, because an input file uploaded under the
wrong purpose is rejected at creation and never becomes a batch object at all.
"""
import argparse
import logging
import os
import sys
import time

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("openai_batch_validation_audit")

BATCHES_URL = "https://api.openai.com/v1/batches"
FILES_URL = "https://api.openai.com/v1/files"

# The batch list takes limit (1-100) and after. There is no status filter and no
# date range, so every bit of selection below happens on this side of the wire.
PAGE = 100

# Purposes the Files API namespaces separately from batch input. A .jsonl parked
# under one of these was uploaded for a batch that /v1/batches refused, since it
# accepts an input file only when purpose is exactly "batch".
NOT_BATCH_INPUT = ("user_data", "assistants", "fine-tune", "vision")

# Error codes the docs and the endpoint's own validation produce often enough to
# be worth a specific repair line rather than the generic one.
KNOWN_CODES = {
    "invalid_json": "a line is not valid JSON. Validate the file locally before "
                    "upload: every line must parse on its own.",
    "duplicate_custom_id": "two lines share a custom_id. They must be unique "
                           "within the file, because results come back unordered "
                           "and custom_id is the only join key.",
    "missing_required_parameter": "a line is missing a required field. Each row "
                                  "needs custom_id, method, url and body.",
    "invalid_url": "a line's url does not match the batch endpoint. The two must "
                   "agree for every row in the file.",
    "model_not_found": "the body names a model this project cannot reach. Check "
                       "the id against GET /v1/models with the same key.",
    "empty_file": "the input file has no lines in it. The upload succeeded and "
                  "the content did not.",
}

FINDINGS = ("validation-failed", "orphan-input-files")


def failed_batches(batches):
    """Batches whose input file was rejected. Pure.

    The only status that means "validation refused this file". A batch that ran
    and had rows fail inside it reports completed with a non-zero
    request_counts.failed, which this script deliberately does not look at.
    """
    return [b for b in batches or [] if (b or {}).get("status") == "failed"]


def error_rows(batch):
    """Normalised entries from errors.data[]. Pure. Never raises on a shape.

    errors is optional, its data list is optional, and an entry's line is
    documented as "if applicable", so every field here is allowed to be absent.
    """
    errors = (batch or {}).get("errors") or {}
    data = errors.get("data") if isinstance(errors, dict) else None
    out = []
    for item in data or []:
        if not isinstance(item, dict):
            continue
        line = item.get("line")
        try:
            line = int(line)
        except (TypeError, ValueError):
            line = None
        out.append({"code": str(item.get("code") or "unknown"),
                    "message": str(item.get("message") or ""),
                    "param": item.get("param"),
                    "line": line})
    return out


def lines_by_code(rows):
    """{code: (sorted lines, count, one message, one param)}. Pure.

    A file with 40,000 bad rows produces 40,000 near-identical messages. One row
    per code carrying the line numbers is the same information and is readable.
    """
    grouped = {}
    for row in rows or []:
        slot = grouped.setdefault(row["code"], {"lines": set(), "count": 0,
                                                "message": "", "param": None})
        slot["count"] += 1
        if row.get("line") is not None:
            slot["lines"].add(row["line"])
        if not slot["message"]:
            slot["message"] = row.get("message") or ""
        if slot["param"] is None:
            slot["param"] = row.get("param")
    return {code: (sorted(v["lines"]), v["count"], v["message"], v["param"])
            for code, v in sorted(grouped.items())}


def nothing_billed(batch):
    """True when the batch dispatched no requests at all. Pure.

    Validation runs before dispatch, so a failed batch has an all-zero
    request_counts. An absent counts object is treated as zero, which is what a
    batch that never started actually looks like.
    """
    counts = (batch or {}).get("request_counts") or {}
    try:
        return all(int(counts.get(k) or 0) == 0
                   for k in ("total", "completed", "failed"))
    except (TypeError, ValueError):
        return False


def batch_input_ids(batches):
    """Every input_file_id the account has ever handed to a batch. Pure."""
    return {str(b.get("input_file_id")) for b in batches or []
            if (b or {}).get("input_file_id")}


def mispurposed_inputs(files, used_ids):
    """.jsonl files that can never be batch input and never were. Pure.

    All three conditions matter. Drop the last one and every input file you ever
    used successfully gets flagged, which is how a check gets switched off.
    """
    out = []
    for f in files or []:
        if not isinstance(f, dict):
            continue
        name = str(f.get("filename") or "")
        purpose = str(f.get("purpose") or "")
        if not name.lower().endswith(".jsonl"):
            continue
        if purpose not in NOT_BATCH_INPUT:
            continue
        if str(f.get("id")) in (used_ids or set()):
            continue
        out.append({"id": str(f.get("id")), "filename": name,
                    "purpose": purpose, "bytes": int(f.get("bytes") or 0)})
    return sorted(out, key=lambda r: r["id"])


def within_window(batch, now, days):
    """True when the batch was created inside the window. Pure. days<=0 is all."""
    if not days or days <= 0:
        return True
    try:
        created = int((batch or {}).get("created_at") or 0)
    except (TypeError, ValueError):
        return False
    return created >= now - days * 86400


def verdict(failed, orphans, days):
    """Grade the run. Pure. Returns (state, detail)."""
    failed = list(failed or [])
    orphans = list(orphans or [])
    window = ("in the last %d days" % days) if days and days > 0 else "in the account"
    if failed and orphans:
        return ("validation-failed",
                "%d batch(es) failed input validation %s, and %d .jsonl was "
                "uploaded under a purpose /v1/batches will not accept"
                % (len(failed), window, len(orphans)))
    if failed:
        return ("validation-failed",
                "%d batch(es) failed input validation %s and nothing polled "
                "them to find out" % (len(failed), window))
    if orphans:
        return ("orphan-input-files",
                "%d .jsonl file(s) sit under a purpose /v1/batches will not "
                "accept, referenced by no batch" % len(orphans))
    return ("validation-clean",
            "no batch %s failed validation, and every .jsonl in the file list "
            "either carries purpose=batch or was used by a batch" % window)


def repair_lines(state, codes):
    """The repair for one verdict. Pure. Printed, never performed."""
    if state == "validation-clean":
        return ["nothing to change. Keep the assertion that a submitter only "
                "logs success once status has left \"validating\"."]
    lines = []
    for code in sorted(set(codes or [])):
        if code in KNOWN_CODES:
            lines.append("%s: %s" % (code, KNOWN_CODES[code]))
    if state == "validation-failed":
        lines.append("fix the input at the reported lines, then re-upload with "
                     "purpose=batch and create the batch again. Nothing was "
                     "billed, so nothing needs reconciling.")
        lines.append("make the submitter poll. A 200 from batch creation is a "
                     "receipt, not a result: the only honest success signal is "
                     "a batch that has left \"validating\".")
    if state == "orphan-input-files":
        lines.append("re-upload each file with purpose=batch and delete the "
                     "mis-purposed copy, which counts against project storage "
                     "until you do.")
        lines.append("assert in the upload helper that the purpose matches the "
                     "endpoint that will consume the file.")
    return lines


def get_json(url, key, params=None, timeout=30):
    """One GET. Returns (payload, error). Read only, always."""
    try:
        r = requests.get(url, headers={"Authorization": "Bearer %s" % key},
                         params=params or {}, timeout=timeout)
    except requests.RequestException as exc:
        return (None, "request failed: %s" % exc)
    if r.status_code != 200:
        detail = ""
        try:
            detail = str((r.json().get("error") or {}).get("message") or "")
        except ValueError:
            detail = (r.text or "")[:160]
        return (None, "HTTP %d %s" % (r.status_code, detail))
    try:
        return (r.json(), None)
    except ValueError:
        return (None, "response was not JSON")


def page_all(url, key, params, max_pages):
    """Follow the after cursor. Returns (rows, error). GETs only."""
    rows = []
    after = None
    for _ in range(max(1, max_pages)):
        query = dict(params or {})
        if after:
            query["after"] = after
        payload, err = get_json(url, key, query)
        if err:
            return (rows, err)
        data = payload.get("data") or []
        rows.extend(data)
        if not payload.get("has_more") or not data:
            break
        after = data[-1].get("id")
        if not after:
            break
    return (rows, None)


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--since-days", type=int, default=30,
                    help="only report batches created inside this window (0 = all)")
    ap.add_argument("--max-pages", type=int, default=20,
                    help="cap on pages of 100 for each list; a bounded read "
                         "gives a bounded answer and the output says so")
    args = ap.parse_args()

    key = os.environ.get("OPENAI_API_KEY")
    if not key:
        log.error("set OPENAI_API_KEY to a project key set to Read Only")
        return 2

    now = int(time.time())
    batches, err = page_all(BATCHES_URL, key, {"limit": PAGE}, args.max_pages)
    if err and not batches:
        log.error("could not read the batch list: %s", err)
        return 2
    if err:
        log.warning("batch list stopped early: %s", err)

    scoped = [b for b in batches if within_window(b, now, args.since_days)]
    failed = failed_batches(scoped)
    seen_codes = []
    for b in failed:
        billed = "nothing billed (0 requests)" if nothing_billed(b) \
            else "request_counts is not all zero, which is unusual for failed"
        stamp = time.strftime("%Y-%m-%dT%H:%M:%SZ",
                              time.gmtime(int(b.get("failed_at")
                                              or b.get("created_at") or 0)))
        log.warning("%-16s failed at %s  %s", b.get("id"), stamp, billed)
        groups = lines_by_code(error_rows(b))
        if not groups:
            log.warning("  (the errors object is empty, so the reason is not "
                        "readable from the API)")
        for code, (lines, count, message, param) in groups.items():
            seen_codes.append(code)
            shown = ", ".join(str(n) for n in lines[:6])
            more = " and %d more" % (len(lines) - 6) if len(lines) > 6 else ""
            where = ("lines %s%s" % (shown, more)) if lines else "no line given"
            extra = "  param %s" % param if param else ""
            log.warning("  %-26s %s%s", code, where, extra)
            if message:
                log.info("  %-26s %s", "", message[:140])

    files, ferr = page_all(FILES_URL, key, {"limit": 10000}, args.max_pages)
    if ferr:
        log.warning("file list stopped early: %s", ferr)
    orphans = mispurposed_inputs(files, batch_input_ids(batches))
    for row in orphans:
        log.warning("orphan-input    %s  %s  purpose=%s  %.1f MB", row["id"],
                    row["filename"], row["purpose"], row["bytes"] / 1048576.0)

    state, detail = verdict(failed, orphans, args.since_days)
    emit = log.warning if state in FINDINGS else log.info
    emit("%-20s %s", state, detail)
    emit("  measured: status, errors.data[] and request_counts from the batch "
         "list, purpose from the file list")
    emit("  inferred: that the pipeline never polled, since a failed batch is "
         "otherwise indistinguishable from one nobody re-ran on purpose")
    for line in repair_lines(state, seen_codes):
        emit("  repair: %s", line)

    total = len(failed) + len(orphans)
    log.info("%d finding(s)", total)
    return 1 if total else 0


if __name__ == "__main__":
    sys.exit(main())

File: batch-failed-input-validation/node/openai-batch-validation-audit.mjs
/**
 * Report OpenAI batches that failed input validation, and the lines that broke.
 *
 * Read only. Two GET endpoints, /v1/batches and /v1/files. No upload, no batch
 * creation, no re-submission of a failed batch.
 *
 * Status "failed" on this API means the input file did not survive validation,
 * which happens before any request reaches the model: request_counts is all
 * zeros and nothing was billed. Rows that failed inside a batch that ran are a
 * different status and a different note.
 */
const BATCHES_URL = 'https://api.openai.com/v1/batches';
const FILES_URL = 'https://api.openai.com/v1/files';

const PAGE = 100;

export const NOT_BATCH_INPUT = new Set(['user_data', 'assistants', 'fine-tune', 'vision']);

export const KNOWN_CODES = {
  invalid_json: 'a line is not valid JSON. Validate the file locally before '
    + 'upload: every line must parse on its own.',
  duplicate_custom_id: 'two lines share a custom_id. They must be unique within '
    + 'the file, because results come back unordered and custom_id is the only '
    + 'join key.',
  missing_required_parameter: 'a line is missing a required field. Each row needs '
    + 'custom_id, method, url and body.',
  invalid_url: "a line's url does not match the batch endpoint. The two must "
    + 'agree for every row in the file.',
  model_not_found: 'the body names a model this project cannot reach. Check the '
    + 'id against GET /v1/models with the same key.',
  empty_file: 'the input file has no lines in it. The upload succeeded and the '
    + 'content did not.',
};

const FINDINGS = new Set(['validation-failed', 'orphan-input-files']);

/** Batches whose input file was rejected. Pure. */
export function failedBatches(batches) {
  return (batches ?? []).filter((b) => (b ?? {}).status === 'failed');
}

/** Normalised entries from errors.data[]. Pure. Never throws on a shape. */
export function errorRows(batch) {
  const errors = (batch ?? {}).errors;
  const data = errors && typeof errors === 'object' ? errors.data : null;
  const out = [];
  for (const item of data ?? []) {
    if (!item || typeof item !== 'object') continue;
    const parsed = Number.parseInt(item.line, 10);
    out.push({
      code: String(item.code ?? 'unknown'),
      message: String(item.message ?? ''),
      param: item.param ?? null,
      line: Number.isFinite(parsed) ? parsed : null,
    });
  }
  return out;
}

/** {code: [lines, count, message, param]}. Pure. Sorted by code, then line. */
export function linesByCode(rows) {
  const grouped = new Map();
  for (const row of rows ?? []) {
    if (!grouped.has(row.code)) {
      grouped.set(row.code, { lines: new Set(), count: 0, message: '', param: null });
    }
    const slot = grouped.get(row.code);
    slot.count += 1;
    if (row.line !== null && row.line !== undefined) slot.lines.add(row.line);
    if (!slot.message) slot.message = row.message ?? '';
    if (slot.param === null) slot.param = row.param ?? null;
  }
  const out = {};
  for (const code of [...grouped.keys()].sort()) {
    const slot = grouped.get(code);
    out[code] = [[...slot.lines].sort((a, b) => a - b), slot.count, slot.message,
                 slot.param];
  }
  return out;
}

/** True when the batch dispatched no requests at all. Pure. */
export function nothingBilled(batch) {
  const counts = (batch ?? {}).request_counts ?? {};
  return ['total', 'completed', 'failed']
    .every((k) => (Number(counts[k]) || 0) === 0);
}

/** Every input_file_id the account has handed to a batch. Pure. */
export function batchInputIds(batches) {
  const out = new Set();
  for (const b of batches ?? []) {
    if ((b ?? {}).input_file_id) out.add(String(b.input_file_id));
  }
  return out;
}

/** .jsonl files that can never be batch input and never were. Pure. */
export function mispurposedInputs(files, usedIds) {
  const used = usedIds ?? new Set();
  return (files ?? [])
    .filter((f) => f && typeof f === 'object')
    .filter((f) => String(f.filename ?? '').toLowerCase().endsWith('.jsonl'))
    .filter((f) => NOT_BATCH_INPUT.has(String(f.purpose ?? '')))
    .filter((f) => !used.has(String(f.id)))
    .map((f) => ({ id: String(f.id), filename: String(f.filename ?? ''),
                   purpose: String(f.purpose ?? ''), bytes: Number(f.bytes) || 0 }))
    .sort((a, b) => a.id.localeCompare(b.id));
}

/** True when the batch was created inside the window. Pure. */
export function withinWindow(batch, now, days) {
  if (!days || days <= 0) return true;
  const created = Number((batch ?? {}).created_at) || 0;
  return created >= now - days * 86400;
}

/** Grade the run. Pure. Returns [state, detail]. */
export function verdict(failed, orphans, days) {
  const f = (failed ?? []).length;
  const o = (orphans ?? []).length;
  const window = days && days > 0 ? `in the last ${days} days` : 'in the account';
  if (f && o) {
    return ['validation-failed',
      `${f} batch(es) failed input validation ${window}, and ${o} .jsonl was `
      + 'uploaded under a purpose /v1/batches will not accept'];
  }
  if (f) {
    return ['validation-failed',
      `${f} batch(es) failed input validation ${window} and nothing polled them `
      + 'to find out'];
  }
  if (o) {
    return ['orphan-input-files',
      `${o} .jsonl file(s) sit under a purpose /v1/batches will not accept, `
      + 'referenced by no batch'];
  }
  return ['validation-clean',
    `no batch ${window} failed validation, and every .jsonl in the file list `
    + 'either carries purpose=batch or was used by a batch'];
}

/** The repair for one verdict. Pure. Printed, never performed. */
export function repairLines(state, codes) {
  if (state === 'validation-clean') {
    return ['nothing to change. Keep the assertion that a submitter only logs '
      + 'success once status has left "validating".'];
  }
  const lines = [];
  for (const code of [...new Set(codes ?? [])].sort()) {
    if (KNOWN_CODES[code]) lines.push(`${code}: ${KNOWN_CODES[code]}`);
  }
  if (state === 'validation-failed') {
    lines.push('fix the input at the reported lines, then re-upload with '
      + 'purpose=batch and create the batch again. Nothing was billed, so '
      + 'nothing needs reconciling.');
    lines.push('make the submitter poll. A 200 from batch creation is a receipt, '
      + 'not a result: the only honest success signal is a batch that has left '
      + '"validating".');
  }
  if (state === 'orphan-input-files') {
    lines.push('re-upload each file with purpose=batch and delete the '
      + 'mis-purposed copy, which counts against project storage until you do.');
    lines.push('assert in the upload helper that the purpose matches the '
      + 'endpoint that will consume the file.');
  }
  return lines;
}

async function getJson(url, key, params) {
  const target = new URL(url);
  for (const [k, v] of Object.entries(params ?? {})) {
    target.searchParams.set(k, String(v));
  }
  let res;
  try {
    res = await fetch(target, { headers: { Authorization: `Bearer ${key}` } });
  } catch (err) {
    return [null, `request failed: ${err.message}`];
  }
  if (res.status !== 200) {
    let detail = '';
    try { detail = String((await res.json())?.error?.message ?? ''); } catch { detail = ''; }
    return [null, `HTTP ${res.status} ${detail}`];
  }
  try {
    return [await res.json(), null];
  } catch {
    return [null, 'response was not JSON'];
  }
}

async function pageAll(url, key, params, maxPages) {
  const rows = [];
  let after = null;
  for (let i = 0; i < Math.max(1, maxPages); i += 1) {
    const query = { ...(params ?? {}) };
    if (after) query.after = after;
    const [payload, err] = await getJson(url, key, query);
    if (err) return [rows, err];
    const data = payload.data ?? [];
    rows.push(...data);
    if (!payload.has_more || !data.length) break;
    after = data[data.length - 1]?.id;
    if (!after) break;
  }
  return [rows, null];
}

function args(argv) {
  const out = { sinceDays: 30, maxPages: 20 };
  for (let i = 0; i < argv.length; i += 1) {
    if (argv[i] === '--since-days') out.sinceDays = Number.parseInt(argv[i += 1], 10);
    else if (argv[i] === '--max-pages') out.maxPages = Number.parseInt(argv[i += 1], 10);
  }
  return out;
}

async function main() {
  const opts = args(process.argv.slice(2));
  const key = (process.env.OPENAI_API_KEY || "dummy-openai-api-key");
  if (!key) {
    console.error('set OPENAI_API_KEY to a project key set to Read Only');
    process.exitCode = 2;
    return;
  }

  const now = Math.floor(Date.now() / 1000);
  const [batches, err] = await pageAll(BATCHES_URL, key, { limit: PAGE }, opts.maxPages);
  if (err && !batches.length) {
    console.error(`could not read the batch list: ${err}`);
    process.exitCode = 2;
    return;
  }
  if (err) console.log(`batch list stopped early: ${err}`);

  const scoped = batches.filter((b) => withinWindow(b, now, opts.sinceDays));
  const failed = failedBatches(scoped);
  const seenCodes = [];
  for (const b of failed) {
    const billed = nothingBilled(b)
      ? 'nothing billed (0 requests)'
      : 'request_counts is not all zero, which is unusual for failed';
    const stamp = new Date((Number(b.failed_at || b.created_at) || 0) * 1000)
      .toISOString().replace(/\.\d+Z$/, 'Z');
    console.log(`${String(b.id).padEnd(16)} failed at ${stamp}  ${billed}`);
    const groups = linesByCode(errorRows(b));
    if (!Object.keys(groups).length) {
      console.log('  (the errors object is empty, so the reason is not readable '
        + 'from the API)');
    }
    for (const [code, [lines, , message, param]] of Object.entries(groups)) {
      seenCodes.push(code);
      const shown = lines.slice(0, 6).join(', ');
      const more = lines.length > 6 ? ` and ${lines.length - 6} more` : '';
      const where = lines.length ? `lines ${shown}${more}` : 'no line given';
      const extra = param ? `  param ${param}` : '';
      console.log(`  ${code.padEnd(26)} ${where}${extra}`);
      if (message) console.log(`  ${''.padEnd(26)} ${message.slice(0, 140)}`);
    }
  }

  const [files, ferr] = await pageAll(FILES_URL, key, { limit: 10000 }, opts.maxPages);
  if (ferr) console.log(`file list stopped early: ${ferr}`);
  const orphans = mispurposedInputs(files, batchInputIds(batches));
  for (const row of orphans) {
    console.log(`orphan-input    ${row.id}  ${row.filename}  purpose=${row.purpose}  `
      + `${(row.bytes / 1048576).toFixed(1)} MB`);
  }

  const [state, detail] = verdict(failed, orphans, opts.sinceDays);
  console.log(`${state.padEnd(20)} ${detail}`);
  console.log('  measured: status, errors.data[] and request_counts from the '
    + 'batch list, purpose from the file list');
  console.log('  inferred: that the pipeline never polled, since a failed batch '
    + 'is otherwise indistinguishable from one nobody re-ran on purpose');
  for (const line of repairLines(state, seenCodes)) console.log(`  repair: ${line}`);

  const total = failed.length + orphans.length;
  console.log(`${total} finding(s)`);
  process.exitCode = total ? 1 : 0;
  void FINDINGS;
}

if (import.meta.url === `file://${process.argv[1]}`) await main();
